File: modules/analysis/templates/03-regression/00-plan-dispatch/plan_dispatch.py
#!/usr/bin/env python3
"""Read analysis-execution-plan and create regression dispatch files."""

from __future__ import annotations

import argparse
import csv
import re
import sys
from datetime import date
from pathlib import Path


def _load_common():
    for parent in Path(__file__).resolve().parents:
        shared = parent / "_shared"
        if shared.exists():
            sys.path.insert(0, str(shared))
            import python_common  # type: ignore

            return python_common
    raise RuntimeError("Cannot locate templates/_shared/python_common.py")


common = _load_common()

TASK_RULES = {
    "main": ["Y", "X", "control", "fe", "cluster", "weight", "主回归", "描述统计"],
    "nonlinear": ["nonlinear", "logit", "probit", "poisson", "tobit", "heckman", "有序", "多分类"],
    "panel": ["id", "time", "panel", "FE", "RE", "Hausman", "面板"],
    "causal": ["instrument", "treatment", "post", "running", "cutoff", "IV", "DID", "RDD", "PSM", "CEM", "IPW", "SCM"],
    "mechanism_heterogeneity": ["mediator", "mechanism", "moderator", "heterogeneity", "threshold", "中介", "机制", "调节", "异质性", "门槛", "交互", "分组"],
    "robustness": ["robustness", "placebo", "sensitivity", "替代变量", "替代样本", "替代模型", "稳健性", "安慰剂"],
}


def infer_tasks(text: str) -> list[dict[str, str]]:
    rows = []
    for task, needles in TASK_RULES.items():
        matched = sorted({n for n in needles if re.search(re.escape(n), text, flags=re.IGNORECASE)})
        rows.append(
            {
                "task": task,
                "run": "yes" if matched else "blocked",
                "evidence": "; ".join(matched) if matched else "analysis-execution-plan 未发现触发角色",
                "blocker": "" if matched else "缺少变量角色或设计信号；回流变量发现/清洗复核",
            }
        )
    return rows


def main() -> int:
    parser = common.add_common_args(argparse.ArgumentParser(description=__doc__))
    args = parser.parse_args()
    p = common.paths(args.out_root)
    today = date.today().isoformat()
    plan = Path(args.plan) if args.plan else common.find_latest_report(args.out_root, "analysis-execution-plan")
    plan_text = plan.read_text(encoding="utf-8", errors="ignore") if plan and plan.exists() else ""
    dict_path = Path(args.dict_path) if args.dict_path else p["data"] / "variable-dictionary.csv"
    dict_rows = common.read_csv_dicts(dict_path)
    cleaning = common.find_latest_report(args.out_root, "cleaning-report")
    rows = infer_tasks(plan_text)
    json_path = p["reports"] / "regression-dispatch.json"
    csv_path = p["reports"] / "regression-dispatch.csv"
    decision = p["reports"] / f"model-decision-{today}.md"
    common.write_json(
        json_path,
        {
            "plan": str(plan) if plan else None,
            "variable_dictionary": str(dict_path),
            "variable_count": len(dict_rows),
            "cleaning_report": str(cleaning) if cleaning else None,
            "tasks": rows,
        },
    )
    common.write_csv(csv_path, rows, ["task", "run", "evidence", "blocker"])
    common.write_markdown(
        decision,
        "Model Decision",
        {
            "来源计划": str(plan) if plan else "未找到 analysis-execution-plan-[date].md",
            "变量字典与清洗报告": f"变量字典行数：{len(dict_rows)}\n\n清洗报告：{cleaning or '未找到 cleaning-report-[date].md'}",
            "模型派发": "\n".join(f"- {r['task']}: {r['run']} ({r['evidence']})" for r in rows),
            "阻断": "\n".join(f"- {r['task']}: {r['blocker']}" for r in rows if r["blocker"]) or "无派发阻断。",
        },
    )
    status = "ok" if plan_text else "blocked"
    note = "不运行统计模型，只生成 dispatch。" if plan_text else "缺少 analysis-execution-plan；不得直接建模。"
    common.log_run(args, "03-regression/00-plan-dispatch", status, [json_path, csv_path, decision], note)
    return 0 if plan_text else 2


if __name__ == "__main__":
    raise SystemExit(main())

File: modules/analysis/templates/03-regression/00-plan-dispatch/plan_dispatch.R
# Read analysis-execution-plan and create regression dispatch files.

`%||%` <- function(a, b) if (is.null(a) || length(a) == 0) b else a
find_shared <- function() {
  cmd <- commandArgs(FALSE)
  file_arg <- sub("^--file=", "", cmd[grep("^--file=", cmd)][1] %||% "")
  p <- if (nzchar(file_arg)) dirname(normalizePath(file_arg, mustWork = FALSE)) else getwd()
  repeat {
    candidate <- file.path(p, "_shared", "r_common.R")
    if (file.exists(candidate)) return(candidate)
    next_p <- dirname(p)
    if (identical(next_p, p)) stop("Cannot locate templates/_shared/r_common.R")
    p <- next_p
  }
}
source(find_shared())

args <- parse_common_args()
ensure_dirs(args$out_root)
plan <- if (!is.na(args$plan) && file.exists(args$plan)) args$plan else list.files(file.path(args$out_root, "reports"), pattern = "^analysis-execution-plan.*\\.md$", full.names = TRUE)
plan <- if (length(plan)) tail(sort(plan), 1) else NA_character_
txt <- if (!is.na(plan) && file.exists(plan)) paste(readLines(plan, warn = FALSE), collapse = "\n") else ""
dict_path <- if (!is.na(args$dict) && file.exists(args$dict)) args$dict else file.path(args$out_root, "data", "variable-dictionary.csv")
dict_n <- if (file.exists(dict_path)) nrow(utils::read.csv(dict_path, stringsAsFactors = FALSE)) else 0
cleaning <- list.files(file.path(args$out_root, "reports"), pattern = "^cleaning-report.*\\.md$", full.names = TRUE)
cleaning <- if (length(cleaning)) tail(sort(cleaning), 1) else NA_character_
rules <- list(
  main = c("Y", "X", "control", "fe", "cluster", "weight", "主回归", "描述统计"),
  nonlinear = c("nonlinear", "logit", "probit", "poisson", "tobit", "heckman", "有序", "多分类"),
  panel = c("id", "time", "panel", "FE", "RE", "Hausman", "面板"),
  causal = c("instrument", "treatment", "post", "running", "cutoff", "IV", "DID", "RDD", "PSM", "CEM", "IPW", "SCM"),
  mechanism_heterogeneity = c("mediator", "mechanism", "moderator", "heterogeneity", "threshold", "中介", "机制", "调节", "异质性", "门槛", "交互", "分组"),
  robustness = c("robustness", "placebo", "sensitivity", "替代变量", "替代样本", "替代模型", "稳健性", "安慰剂")
)
rows <- do.call(rbind, lapply(names(rules), function(task) {
  hit <- rules[[task]][vapply(rules[[task]], grepl, logical(1), x = txt, ignore.case = TRUE, fixed = TRUE)]
  data.frame(task = task, run = ifelse(length(hit), "yes", "blocked"), evidence = ifelse(length(hit), paste(hit, collapse = "; "), "analysis-execution-plan 未发现触发角色"), blocker = ifelse(length(hit), "", "缺少变量角色或设计信号；回流变量发现/清洗复核"))
}))
csv_path <- file.path(args$out_root, "reports", "regression-dispatch.csv")
json_path <- file.path(args$out_root, "reports", "regression-dispatch.json")
decision <- file.path(args$out_root, "reports", paste0("model-decision-", Sys.Date(), ".md"))
write_csv_safe(rows, csv_path)
writeLines(paste0('{"plan":"', plan, '","variable_count":', dict_n, ',"cleaning_report":"', cleaning, '","note":"R 模板生成；JSON 详情可由 Python 版本重写。"}'), json_path)
write_md(decision, "Model Decision", c("来源计划" = ifelse(is.na(plan), "未找到 analysis-execution-plan-[date].md", plan), "变量字典与清洗报告" = paste0("变量字典行数：", dict_n, "\n\n清洗报告：", ifelse(is.na(cleaning), "未找到 cleaning-report-[date].md", cleaning)), "模型派发" = paste(rows$task, rows$run, rows$evidence, collapse = "\n"), "阻断" = paste(rows$blocker[rows$blocker != ""], collapse = "\n")))
status <- ifelse(nzchar(txt), "ok", "blocked")
log_run(args, "03-regression/00-plan-dispatch", status, c(csv_path, json_path, decision), "不运行统计模型，只生成 dispatch。")
if (status == "blocked") quit(status = 2)
